app/src/main/java/net/gsantner/markor/ui/SearchOrCustomTextDialogCreator.java

```diff
@@ -266,42 +270,50 @@ public static void showSttContextDialog(Activity activity, List<String> availabl
-            dopt.titleText = isTodoTxtAlternativeNaming(activity) ? R.string.category : R.string.context;
-            dopt.searchHintText = R.string.search;
-            SearchOrCustomTextDialog.showMultiChoiceDialogWithSearchFilterUI(activity, dopt);
-        });
-    }
-
-    private static List<String> filterContains(List<String> values, String text) {
-        for (int i = 0; i < values.size(); i++) {
-            if (!values.get(i).contains(text)) {
-                values.remove(i);
-                i--;
-            }
+    public static void showSttKeySearchDialog(
+            final Activity activity,
+            final CharSequence fullText,
+            final Callback.a1<Spannable> highlighter,
+            final boolean isProjects,
+            final Callback.a1<List<Integer>> userCallback)
+    {
+        SearchOrCustomTextDialog.DialogOptions dopt = new SearchOrCustomTextDialog.DialogOptions();
+        baseConf(activity, dopt);
+        final TodoTxtTask[] allTasks = TodoTxtTask.getAllTasks(fullText);
+        dopt.data = Arrays.asList(isProjects ? TodoTxtTask.getProjects(allTasks) : TodoTxtTask.getContexts(allTasks));
+        if (isTodoTxtAlternativeNaming(activity)) {
+            dopt.titleText = isProjects ? R.string.tag : R.string.category;
+        } else {
+            dopt.titleText = isProjects ? R.string.project : R.string.context;
         }
-        return values;
-    }
-
-    private static List<String> filterEmpty(List<String> data) {
-        for (int i = 0; i < data.size(); i++) {
-            if (data.get(i).trim().isEmpty()) {
-                data.remove(i);
-                i--;
+        dopt.searchHintText = R.string.search_or_custom;
+        dopt.iconsForData = Collections.nCopies(dopt.data.size(), R.drawable.ic_search_black_24dp);
+
+        dopt.multiSelectCallback = keyIndices -> {
+            SearchOrCustomTextDialog.DialogOptions dopt2 = new SearchOrCustomTextDialog.DialogOptions();
+            baseConf(activity, dopt2);
+            final Set<String> searchKeys = new HashSet<>(StringUtils.slice((List<String>) dopt.data, keyIndices));
+            final List<Integer> filteredIndices = StringUtils.filterIndices(Arrays.asList(allTasks), task ->
+                    StringUtils.containsAny(searchKeys, Arrays.asList(isProjects ? task.getProjects() : task.getContexts())));
+            dopt2.data = StringUtils.map(StringUtils.slice(Arrays.asList(allTasks), filteredIndices), TodoTxtTask::getLine);
+            dopt2.titleText = dopt.titleText;
+            dopt2.searchHintText = R.string.search;
+            dopt2.highlighter = highlighter;
+            dopt2.multiSelectCallback = posns -> userCallback.callback(StringUtils.slice(filteredIndices, posns));
+            dopt2.positionCallback = posn -> dopt2.multiSelectCallback.callback(Collections.singletonList(posn));
+            SearchOrCustomTextDialog.showMultiChoiceDialogWithSearchFilterUI(activity, dopt2);
+        };
```

Comment:
De-duplicating code for search contexts and search projects. Now supports multi-select. User can now:

1. Multi-select projects (or contexts)
2. See tasks which contain _any_ of the selected projects
3. Multi select tasks
4. Selected tasks will be highlighted

If a user does not multi-select, the dialog works exactly like the old version.